### DISCUSSION TITLE
rocm-smi bandwidth estimation

### DISCUSSION THREAD

**Discussion by claudiubalogh** [ROCm/ROCm#3854] [Q&A] [answered]
Hello!

I am trying to get rocm-smi -b to display estimated PCIEx bandwidth over the last minute while the GPU is uploading/downloading buffers and executing kernels. However, it always displays zero.

I tried using a Radeon Pro W5700 with rocm v5.2.1 and a Radeon Pro W7700 with v6.1.2 on linux.

Is this metric not working for these GPUs, am I using rocm-smi incorrectly or is it a bug?

Thank you!

**Comment by fluidnumericsJoe** [comment]
You should use the `rocm-bandwidth-test` , not `rocm-smi` to get measurements of sustained bandwidth between CPU-GPU and GPU-GPU.

**Comment by claudiubalogh** [comment]
I apologize for not stating my usecase clearly. I would like to measure estimated PCIEx bandwidth while running my OpenCL application, which is long running and does lots of data transfers from multiple CPU threads.

**Reply by fluidnumericsJoe** [reply]
Then you should use a profiler, like `omniperf` .

**Reply by claudiubalogh** [reply]
Of course, but that doesn't answer my question. Thanks!

**Reply by fluidnumericsJoe** [reply]
Are you asking how to measure the achieved PCIe bandwidth of your application during execution? This requires a profiler, like omniperf or rocprof. There's plenty of documentation on both of these for obtaining hardware events profiles.

Or are you asking how to measure what your system is capable of? This requires micro benchmarking, using a tool like rocm-bandwidth-test.

**Reply by claudiubalogh** [reply]
The question is about rocm-smi itself: "Is this metric not working for these GPUs" <i.e. using rocm-smi> ", am I using rocm-smi incorrectly or is it a bug" <in rocm-smi>"?"

**Reply by fluidnumericsJoe** [reply]
You're using the wrong tool for what you are looking for.

**Reply by claudiubalogh** [reply]
I agree with you. Also the question still stands unanswered.

**Reply by fluidnumericsJoe** [reply]
I suspect that since you're on radeon, you're using the radeon drivers.
From the [amd-smi documentation](https://rocm.docs.amd.com/projects/amdsmi/en/latest/install/install.html#supported-platforms)  (which is the successor to rocm-smi) 

> To run the AMD SMI library, the amdgpu driver and the hsmp driver must be installed. Optionally, libdrm can be installed to query firmware information and hardware IPs.

From the [system requirements documentation](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/system-requirements.html), you're GPUs aren't supported. It's not a bug.

**Reply by claudiubalogh** [reply]
The same problem with a supported GPU: Radeon Pro W7800:

```
e1:/opt/rocm/bin>clinfo | grep Board
  Board name:                                    AMD Radeon PRO W7800
e1:/opt/rocm/bin>rocm-smi -b


============================ ROCm System Management Interface ============================
================================ Measured PCIe Bandwidth =================================
GPU[0]          : get_PCIe_bandwidth, Not supported on the given system
==========================================================================================
================================== End of ROCm SMI Log ===================================

```

**Comment by jamesxu2** [comment]
Hi @claudiubalogh, 
> Is this metric not working for these GPUs, am I using rocm-smi incorrectly or is it a bug?

This is not a bug and you are not using rocm-smi incorrectly - the pcie_bw metric is not supported for the W7800. The way this metric (and many others) work is that the hardware must write this information to a system debug file under /sys/class/drm/card**X**/device/**pcie_bw** ([ref](https://github.com/ROCm/rocm_smi_lib/blob/35c1d00f5a603988974f7148cf0dfd7c28c06f18/src/rocm_smi_device.cc#L102)). Your device does not export this file and therefore cannot report this metric. AMD SMI and ROCm SMI both tap into this file to gather the PCIe bandwidth metric so neither tool would be able to locate it.

As far as I'm aware, there isn't a publicly available table of the supported commands for AMD-SMI and ROCm-SMI organized by hardware. However, the source code is available and our community support team is also happy to answer further questions.

**Reply by claudiubalogh** [reply]
rocm lists the Radeon Pro W7800 since [v6.0.2](https://rocm.docs.amd.com/projects/install-on-linux/en/docs-6.0.2/reference/system-requirements.html) as "✅: Supported - Official software distributions of the current ROCm release fully support this hardware."

So is the statement in the table false or how should we read it?

Are all shortcomings of all fully supported cards documented somewhere?
I don't think it's feasible for everyone to dig into the source code.

**Reply by jamesxu2** [reply, marked as answer]
@claudiubalogh, I recognize that statement about support for the W7800 might be reductive. From a theoretical standpoint, architectural differences between different cards make it impossible for "fully supported" to mean exactly the same thing on different cards; our flagship MI-series accelerators have more features, metrics and optimizations than consumer Radeon cards for example.  

One interpretation of "_Official software distributions of the current ROCm release fully support this hardware_" is that we run extensive testing on the supported card, and expect it to work for most common workloads. There are edgecases like yours - the pcie_bw metric is not exported by most of our cards - but I don't think those shortcomings are fully documented anywhere unfortunately. You can imagine that such a support matrix - with every device in every configuration (eg. baremetal, virtualized, partitioned), and every feature of every component, would be massive and unmaintainable. Here's what we're doing to fill the gap:

1. We try to be verbose where features are not supported within the software itself. The error message "`get_PCIe_bandwidth, Not supported on the given system`" is self explanatory.
2. Our documentation discusses architecture specific limitations in some cases ([example](https://rocm.docs.amd.com/projects/omnitrace/en/main/how-to/configuring-runtime-options.html#exploring-hardware-counters) from our rocprof-sys profiler). It is admittedly sparse, but we are constantly working to improve it. 
3. Our community support team is happy to field any questions you have. We've been investing into expanding our support capability so we can respond to issues quickly or escalate them internally. 
4. As a last resort, most of our code is open source so you could debug the issue yourself, but that's not our expectation.
